codec-http2/src/test/java/io/netty/handler/codec/http2/CleartextHttp2ServerUpgradeHandlerTest.java

```diff
@@ -0,0 +1,176 @@
+
+    @Before
+    public void setUp() {
+        frameListener = mock(Http2FrameListener.class);
+
+        http2ConnectionHandler = new Http2ConnectionHandlerBuilder().frameListener(frameListener).build();
+        http2ConnectionHandler = spy(http2ConnectionHandler);
+
+        upgradeCodecFactory = new UpgradeCodecFactory() {
+            @Override
+            public UpgradeCodec newUpgradeCodec(CharSequence protocol) {
+                return new Http2ServerUpgradeCodec(http2ConnectionHandler);
+            }
+        };
+        upgradeCodecFactory = spy(upgradeCodecFactory);
+
+        HttpServerCodec httpServerCodec = new HttpServerCodec();
+        HttpServerUpgradeHandler upgradeHandler = new HttpServerUpgradeHandler(httpServerCodec, upgradeCodecFactory);
+
+        CleartextHttp2ServerUpgradeHandler handler = new CleartextHttp2ServerUpgradeHandler(
+                httpServerCodec, upgradeHandler, http2ConnectionHandler);
+        channel = new EmbeddedChannel(handler);
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        channel.finishAndReleaseAll();
+    }
+
+    @Test
+    public void priorKnowledge() throws Exception {
+        channel.writeInbound(Http2CodecUtil.connectionPrefaceBuf());
+
+        ByteBuf settingsFrame = settingsFrameBuf();
+
+        assertFalse(channel.writeInbound(settingsFrame));
+
+        InOrder inOrder = inOrder(upgradeCodecFactory, http2ConnectionHandler, frameListener);
+        inOrder.verify(upgradeCodecFactory, never()).newUpgradeCodec(anyString());
+        inOrder.verify(http2ConnectionHandler, times(2)).channelRead(
+                any(ChannelHandlerContext.class), any(ByteBuf.class));
+        inOrder.verify(frameListener).onSettingsRead(
+                any(ChannelHandlerContext.class), eq(expectedSettings()));
+    }
+
+    @Test
+    public void upgrade() throws Exception {
+        String upgradeString = "GET / HTTP/1.1\r\n" +
+                               "Host: example.com\r\n" +
+                               "Connection: Upgrade, HTTP2-Settings\r\n" +
+                               "Upgrade: h2c\r\n" +
+                               "HTTP2-Settings: AAMAAABkAAQAAP__\r\n\r\n";
+        ByteBuf upgrade = Unpooled.buffer().writeBytes(upgradeString.getBytes(CharsetUtil.US_ASCII));
+
+        assertFalse(channel.writeInbound(upgrade));
+
+        InOrder inOrder = inOrder(upgradeCodecFactory, http2ConnectionHandler);
+        inOrder.verify(upgradeCodecFactory).newUpgradeCodec("h2c");
+        inOrder.verify(http2ConnectionHandler).onHttpServerUpgrade(
+                new Http2Settings().maxConcurrentStreams(100).initialWindowSize(65535));
+        inOrder.verify(http2ConnectionHandler).userEventTriggered(
+                any(ChannelHandlerContext.class), isA(UpgradeEvent.class));
+    }
+
+    @Test
+    public void priorKnowledgeInFragments() throws Exception {
+        ByteBuf connectionPreface = Http2CodecUtil.connectionPrefaceBuf();
```

Comment:
`Http2CodecUtil.connectionPrefaceBuf()` will return a retained slice ... it seems like the pipeline is not hooked up correctly because the reference count doesn't get decremented at the end of this test ... lets re-evaluate once the `spy` calls have been removed.